File: src/main/java/cn/jackuxl/qforum/controller/ThreadController.java
package cn.jackuxl.qforum.controller;

import cn.jackuxl.qforum.model.Thread;
import cn.jackuxl.qforum.model.User;
import cn.jackuxl.qforum.serviceimpl.BoardServiceImpl;
import cn.jackuxl.qforum.serviceimpl.ThreadServiceImpl;
import cn.jackuxl.qforum.serviceimpl.UserServiceImpl;
import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletResponse;
import java.util.List;

@CrossOrigin
@RestController
public class ThreadController {
    @Autowired
    private UserServiceImpl userService;
    @Autowired
    private ThreadServiceImpl threadService;
    @Autowired
    private BoardServiceImpl boardService;
    @Autowired
    private HttpServletResponse response;

    @RequestMapping(value = "/thread/post", produces = "application/json;charset=UTF-8")
    public String postThread(String sessionId, Thread thread,int boardId) {
        JSONObject result = new JSONObject();
        User user = userService.getUserBySessionId(sessionId);
        if (user != null && sessionId != null) {
            if(boardService.getBoardById(boardId)==null){
                result.put("code", 403);
                result.put("error", "no_such_board");
            }
            else if(thread.getTitle().isEmpty()){
                result.put("code", 403);
                result.put("error", "title_cannot_be_empty");
            }
            else{
                thread.setPublisherId(user.getId());
                thread.setPostTime(String.valueOf(System.currentTimeMillis()));
                thread.setBoardId(boardId);
                if (threadService.postThread(thread) > 0) {
                    result.put("code", 200);
                    result.put("msg", "success");
                    List<Thread> threads = threadService.listThreads(boardId);
                    result.put("id",threads.get(threads.size()-1).getId());
                } else {
                    result.put("code", 403);
                    result.put("error", "unknown");
                }
            }
        } else {
            result.put("code", 403);
            result.put("error", "no_such_user");
        }
        response.setStatus(result.getInteger("code"));
        return result.toJSONString();
    }

    @RequestMapping(value = "/thread/getThreadDetail", produces = "application/json;charset=UTF-8")
    public String getThreadDetail(int id) {
        JSONObject result = new JSONObject();
        Thread thread  = threadService.getThreadById(id);
        if (thread!=null) {
            result = JSON.parseObject(JSON.toJSONString(thread));
            result.put("code",200);
            result.put("publisher",getPublicPublisherObject(result.getInteger("publisherId")));
            result.remove("publisherId");
        } else {
            result.put("code", 403);
            result.put("error", "no_such_thread");
        }
        response.setStatus(result.getInteger("code"));
        return result.toJSONString();
    }

    @RequestMapping(value = "/thread/list", produces = "application/json;charset=UTF-8")
    public String listThreads(int boardId) {
        JSONObject result = new JSONObject();
        if(boardService.getBoardById(boardId)!=null){
            result.put("code", 200);
            result.put("msg", "success");
            List<Thread> threads = threadService.listThreads(boardId);
            JSONArray tmp = JSON.parseArray(JSON.toJSONString(threads));
            for(int i = 0;i<tmp.size();i++){
                tmp.getJSONObject(i).put("publisher",getPublicPublisherObject(tmp.getJSONObject(i).getInteger("publisherId")));
                tmp.getJSONObject(i).remove("publisherId");
            }
            result.put("threadList",tmp);
            result.put("size",threads.size());

        }
        else{
            result.put("code", 403);
            result.put("error", "no_such_board");
        }
        response.setStatus(result.getInteger("code"));
        return result.toJSONString();
    }

    private JSONObject getPublicPublisherObject(int uid){
        JSONObject publisher = JSON.parseObject(JSON.toJSONString(userService.getUserById(uid)));
        publisher.remove("password");
        publisher.remove("lastLoginIp");
        publisher.remove("salt");
        publisher.remove("sessionId");
        return publisher;
    }
}

File: src/main/java/cn/jackuxl/qforum/controller/ThreadController.kt
package cn.jackuxl.qforum.controller

import cn.dev33.satoken.stp.StpUtil
import cn.jackuxl.qforum.constants.StaticProperty
import cn.jackuxl.qforum.entity.Thread
import cn.jackuxl.qforum.exception.CustomException
import cn.jackuxl.qforum.model.Result
import cn.jackuxl.qforum.model.ResultEntity
import cn.jackuxl.qforum.service.serviceimpl.BoardServiceImpl
import cn.jackuxl.qforum.service.serviceimpl.ThreadServiceImpl
import cn.jackuxl.qforum.service.serviceimpl.UserServiceImpl
import cn.jackuxl.qforum.util.BasicUtil
import cn.jackuxl.qforum.vo.ThreadVo
import cn.jackuxl.qforum.vo.UserVo
import org.springframework.beans.BeanUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.lang.NonNull
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@CrossOrigin
@RestController
@RequestMapping(value = ["/thread/"], produces = ["application/json;charset=UTF-8"])
class ThreadController {
    @Autowired
    lateinit var userService: UserServiceImpl

    @Autowired
    lateinit var threadService: ThreadServiceImpl

    @Autowired
    lateinit var boardService: BoardServiceImpl

    @RequestMapping(value = ["post"], produces = ["application/json;charset=UTF-8"])
    fun postThread(thread: Thread): ResultEntity<Thread> {
        BasicUtil.assertTool(StpUtil.isLogin() && StpUtil.getLoginId() != null, StaticProperty.NO_SUCH_USER)

        BasicUtil.assertTool(boardService.getBoardById(thread.boardId) != null, StaticProperty.NO_SUCH_THREAD)
        BasicUtil.assertTool(!thread.title.isNullOrBlank(), "title_cannot_be_empty")

        thread.publisherId = StpUtil.getLoginIdAsInt()
        thread.postTime = System.currentTimeMillis().toString()
        thread.likeList = "[]"

        BasicUtil.assertTool(threadService.postThread(thread) > 0, StaticProperty.UNKNOWN)

        val threads = threadService.listThreads(thread.boardId)
        return Result.ok(StaticProperty.SUCCESS, threads[threads.size - 1])
    }

    @RequestMapping(value = ["getThreadDetail"], produces = ["application/json;charset=UTF-8"])
    fun getThreadDetail(id: Int): ResultEntity<ThreadVo> {
        val thread = threadService.getThreadById(id)
        BasicUtil.assertTool(thread != null, StaticProperty.NO_SUCH_THREAD)

        val publisher = UserVo()
        BeanUtils.copyProperties(userService.getUserById(thread.publisherId), publisher)

        val data = ThreadVo(publisher = publisher, board = boardService.getBoardById(thread.boardId))
        BeanUtils.copyProperties(thread, data)

        if (data.likeList == "null") {
            data.likeList = "[]"
        }
        return Result.ok(StaticProperty.SUCCESS, data)
    }

    @RequestMapping(value = ["list"], produces = ["application/json;charset=UTF-8"])
    fun listThreads(boardId: Int?): ResultEntity<MutableList<ThreadVo>> {
        BasicUtil.assertTool(boardService.getBoardById(boardId ?: 0) != null, StaticProperty.NO_SUCH_BOARD)

        val threads = threadService.listThreads(boardId!!)
        val data = mutableListOf<ThreadVo>()
        for (i in threads.indices) {
            val publisher = UserVo()
            BeanUtils.copyProperties(userService.getUserById(threads[i].publisherId), publisher)

            val thread = ThreadVo(publisher = publisher, board = boardService.getBoardById(threads[i].boardId))
            BeanUtils.copyProperties(threads[i], thread)
            data.add(thread)
        }

        return Result.ok(StaticProperty.SUCCESS, data)
    }

    val LIKE = 0
    val DISLIKE = 1

    @RequestMapping(value = ["like"], produces = ["application/json;charset=UTF-8"])
    fun likeThread(@NonNull type: Int, @NonNull tid: Int): ResultEntity<String?> {
        BasicUtil.assertTool(StpUtil.isLogin() && StpUtil.getLoginId() != null, StaticProperty.NO_SUCH_USER)

        when (type) {
            LIKE -> {
                BasicUtil.assertTool(
                    threadService.likeThread(tid, StpUtil.getLoginIdAsInt()) > 0,
                    StaticProperty.UNKNOWN
                )
            }
            DISLIKE -> {
                BasicUtil.assertTool(
                    threadService.disLikeThread(tid, StpUtil.getLoginIdAsInt()) > 0,
                    StaticProperty.UNKNOWN
                )
            }
            else -> {
                throw CustomException(msg = StaticProperty.ERROR_TYPE)
            }
        }

        return Result.ok(StaticProperty.SUCCESS)
    }
}